### PR TITLE
Unify `run-tests` and default check handling of `location` field

### PR DESCRIPTION
**Pull request by jackfirth**
This adds a private `location-info` struct alongside `string-info` from #27 to encapsulate rendering a source location with a relative directory path. This obsoletes the logic in both `rackunit/private/format` and `rackunit/text-ui` for handling a check info with key `'location` specially. This is not technically backwards compatible as it changes the observable behavior of `make-check-location`, but it does so in a way that's highly unlikely to affect existing users. See the message of commit 112d4b3 for details.

Works towards closing #35

### CODE REVIEW

**Comment by jeapostrophe** [comment]
I think this is good and should be merged. Any other opinions?

File: rackunit-lib/rackunit/private/check-info.rkt [hunk 1]
```diff
@@ -19,14 +20,22 @@
...
+  (cond
+    [(string-info? info-value) (string-info-value info-value)]
+    [(location-info? info-value)
+     (trim-current-directory
+      (location->string (location-info-value info-value)))]
+    [else (with-output-to-string (λ () (write info-value)))]))
+
+(define (trim-current-directory path)
+  (define cd (path->string (current-directory)))
+  (regexp-replace (regexp-quote cd) path ""))
```
Comment (bennn) [inline rackunit-lib/rackunit/private/check-info.rkt]: I think `(path->string (find-relative-path (current-directory) path))` will have the same result

File: rackunit-test/tests/rackunit/text-ui-util-test.rkt [hunk 2]
```diff
@@ -1,55 +0,0 @@
...
-;; Commentary:
-#lang racket/base
-
-(require rackunit
-         rackunit/private/text-ui-util)
-
-(provide text-ui-util-tests)
-
-(define text-ui-util-tests
-  (test-suite
```
Comment (bennn) [inline rackunit-test/tests/rackunit/text-ui-util-test.rkt]: Please add these to `private/check-info.rkt`

File: rackunit-lib/rackunit/private/check-info.rkt [hunk 3]
```diff
@@ -19,14 +20,22 @@
...
 (define (info-value->string info-value)
-  (if (string-info? info-value)
-      (string-info-value info-value)
-      (with-output-to-string (λ () (write info-value)))))
+  (cond
+    [(string-info? info-value) (string-info-value info-value)]
+    [(location-info? info-value)
+     (trim-current-directory
+      (location->string (location-info-value info-value)))]
+    [else (with-output-to-string (λ () (write info-value)))]))
```
Comment (bennn) [inline rackunit-lib/rackunit/private/check-info.rkt]: Oh hey, this could just be `(~s info-value)` right?
(Maybe `~.s`)

File: rackunit-lib/rackunit/private/check-info.rkt (hunk 1 above)
Comment (jackfirth) [inline rackunit-lib/rackunit/private/check-info.rkt]: Good note, though I'd want to do that in a separate PR.

File: rackunit-lib/rackunit/private/check-info.rkt (hunk 3 above)
Comment (jackfirth) [inline rackunit-lib/rackunit/private/check-info.rkt]: I think `~s` is more appropriate than `~.s` to preserve existing behavior (also chopping off check info seems... weird).

File: rackunit-test/tests/rackunit/text-ui-util-test.rkt (hunk 2 above)
Comment (jackfirth) [inline rackunit-test/tests/rackunit/text-ui-util-test.rkt]: Done

**Comment by AlexKnauth** [comment]
Is it possible that this change broke something in the typed racket tests that only shows up when I run them in DrRacket?

I'm getting an error that seems to come from [rackunit/private/base, line 27](https://github.com/racket/rackunit/blob/master/rackunit-lib/rackunit/private/base.rkt#L27):
```
(define loc (location-info-value (check-info-value maybe-location)))
```
Where `maybe-location` is `(check-info 'location '("<pkgs>/typed-racket-test/unit-tests/typecheck-tests.rkt" 1436 8 60243 109))`, and `(check-info-value maybe-location)` is `'("<pkgs>/typed-racket-test/unit-tests/typecheck-tests.rkt" 1436 8 60243 109)`, where `location-info-value` is failing because it isn't a `location-value` struct.

I'm only getting this error when I run the TR test file from DrRacket, not when I run it from the command line. (And only when there's a test failure.)

**Comment by AlexKnauth** [comment]
The TR test file uses `with-check-info` instead of going through the `make-check-location`, like this:
```
(with-check-info
  (['location (build-source-location-list (quote-srcloc #,stx))])
  ...)
```
Is that not a supported pattern anymore? Should the documentation for `with-check-info` warn about this? Or should `rackunit/private/base` support both patterns?

**Comment by bennn** [comment]
I think `rackunit` should still support that pattern. Using an actual srcloc for the location field should not error, and rackunit should be able to use the srcloc.

**Comment by jackfirth** [comment]
Meant to respond to this earlier, but alas I did not.

Check infos are weird. They're symbol-value mappings with no namespace controls (which identifiers have via scopes and modules) _and_ no value controls (which hashes and other key-value mappings have via contracts). That's a backwards compatibility nightmare, since people will stick who-knows-what in info values and build frameworks and libraries for reading things from specific info keys without attaching contracts.

I'd prefer if `rackunit` handled check infos with the philosophy "no values are special unless wrapped in a generic `rackunit` info struct like `nested-info`, and no keys are special unless a library uses a specific key with a specialized value wrapper struct like `location-info` to do something extra". Anything else invites inconsistency and makes changing things without breaking things hard.

**Comment by AlexKnauth** [comment]
Is that a reason for `(with-check-info* (list (make-check-location ....)) (λ () ....))` and `(with-check-info (['location ....]) ....)` being different?
